docs/rfcs/20221018_cluster_bootstrap.md

```diff
@@ -0,0 +1,325 @@
+
+
+## **Design**
+
+
+### **High level design**
+
+Since there are several issues being addressed in this project, there are a few disparate tasks involved in this overhaul of bootstrapping:
+
+
+
+* Drop the requirement that each node configuration must include a node ID. Users may still supply node IDs, but as a part of bootstrapping a cluster, Redpanda will enforce that newly added nodes will have unique node IDs. Nodes will assign themselves a UUID (persisted in their kv-store), and register themselves with the controller leader, who assigns a unique node ID. Reusing a node ID that has already been used will be strictly disallowed.
+* Remove the period of cluster formation where the cluster only has a single node. Instead, clusters will start with at least three (configurable) nodes, instantiating a controller Raft group with the seed servers list, and growing the cluster as nodes start up.
+* Initial cluster configuration will be done via either environment variables, or bootstrap config file, depending on sensitivity of the configs (e.g. user credentials are secrets, licenses aren’t)
+* Optionally gate the starting of a cluster on the run of a new CLI command. Admins can provide initial user credentials and a license to this command. Before the successful completion of this command, no node in the cluster will respond to Kafka requests, admin requests, etc. This waiting is more of a cherry-on-top as a more sure-fire gate to cluster formation accidents – the above three design points get us most of the way there in solving customer pain points, and will be the focus for Q3.
+
+
+### **Multi-node cluster formation**
+
+Rather than relying on a single node to be the root node, the entire seed servers list will be taken to be the initial Raft group for the controller. Once all the seed servers are up, each seed server should determine whether or not all seed servers have the same list of seed servers. If so, the nodes are primed for starting a cluster and should await instruction to do so. While waiting for start up, we may initialize most subsystems, but must not initialize the controller, a bulk of the admin server, Kafka API, to name a few.
+
+The initialization on seed servers (non-seed servers continue with today’s join-until-success logic), in short:
+
+
+
+* **Seeds:** tries to connect to all seed servers to ensure they all think the list of seed servers, and other cluster configs, is identical. Once this is verified, and no cluster UUID already exists, awaits user input to start its Raft instances with the seed servers list as the initial Raft group.
+* **Non-seeds:** tries to connect to all seed servers to try to join, sending a UUID along with a node ID. No strong requirements on the seed servers list here, other than pointing to a member of the cluster.
+* **Seeds after cluster formation:** as indicated by the presence of a non-empty controller log or a cluster UUID, follows the above non-seeds logic.
+
```

Comment:
I have plant-umled a diagram from my records about that, just in case it can be useful here.

![](https://www.plantuml.com/plantuml/svg/bPCnRiCm34LtdUAF2T2XipCqM_Vs08LgKOqAMZGakgblNqI9ZHWSWd8mR9FyVEAtZrLzqMbWrswSIn6xdgUgL42_gMe5I--PvvSzz4GPkvdgtW50zqOZ-nczP2t2J0KiNqzeoQQ7ugC-WuPHvyFjjOoEk19sMQuozoEZJ3bGiREMmaeKAa99FvJnAQ9LYn_N-aLpYUxz3kESz8GYeeUrff6qATksLrBSc3vIZe9OP81dHWDO1LOkd-UMmidNjNEQKh17iQrjXP5Jhv28a6AaGbatEjK8cUYYJtmmgMNj9h6nmQeP1vkpDhubvGlQWliVWXnIT8-OB4-pScE54RHFtVgv8DFPqZP2o_1m5_RyLh61NA_kI3bS_hW_)

Reply:
One thing I'd change here is that `cluster founder?` should be `is seed?`

It's possible we're a seed server but aren't a founder because a remote cluster UUID exists.